objects/src/accounts/builder/mod.rs

```diff
@@ -0,0 +1,313 @@
+    pub fn new() -> Self {
+        Self {
+            assets: vec![],
+            components: vec![],
+            nonce: ZERO,
+            init_seed: None,
+            account_type: AccountType::RegularAccountUpdatableCode,
+            storage_mode: AccountStorageMode::Private,
+        }
+    }
+
+    /// Sets the initial seed from which the grind for an [`AccountId`] will start. This initial
+    /// seed should come from a cryptographic random number generator.
+    ///
+    ///  This method **must** be called.
+    pub fn init_seed(mut self, init_seed: [u8; 32]) -> Self {
+        self.init_seed = Some(init_seed);
+        self
+    }
+
+    /// Sets the type of the account.
+    pub fn account_type(mut self, account_type: AccountType) -> Self {
+        self.account_type = account_type;
+        self
+    }
+
+    /// Sets the storage mode of the account.
+    pub fn storage_mode(mut self, storage_mode: AccountStorageMode) -> Self {
+        self.storage_mode = storage_mode;
+        self
+    }
+
+    /// Sets the nonce of the account. This method is optional.
+    ///
+    /// If unset, the nonce will default to [`ZERO`].
+    pub fn nonce(mut self, nonce: Felt) -> Self {
+        self.nonce = nonce;
+        self
+    }
+
+    /// Adds the asset to the account's [`AssetVault`]. This method is optional.
+    pub fn with_asset(mut self, asset: Asset) -> Self {
+        self.assets.push(asset);
+        self
+    }
+
+    /// Adds all the assets to the account's [`AssetVault`]. This method is optional.
+    pub fn with_assets<I: IntoIterator<Item = Asset>>(mut self, assets: I) -> Self {
+        self.assets.extend(assets);
+        self
+    }
+
+    /// Adds an [`AccountComponent`] to the builder. This method can be called multiple times and
+    /// **must be called at least once** since an account must export at least one procedure.
+    ///
+    /// All components will be merged to form the final code and storage of the built account.
+    pub fn with_component(mut self, account_component: impl Into<AccountComponent>) -> Self {
+        self.components.push(account_component.into());
+        self
+    }
+
+    /// Builds an [`Account`] out of the configured builder.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - The init seed is not set.
+    /// - If a duplicate assets was added to the builder.
+    /// - Any of the components does not support the set account type.
+    /// - The number of procedures in all merged components is 0 or exceeds
+    ///   [`AccountCode::MAX_NUM_PROCEDURES`](crate::accounts::AccountCode::MAX_NUM_PROCEDURES).
+    /// - Two or more libraries export a procedure with the same MAST root.
+    /// - The number of [`StorageSlot`](crate::accounts::StorageSlot)s of all components exceeds
+    ///   255.
+    /// - [`MastForest::merge`](vm_processor::MastForest::merge) fails on the given components.
+    pub fn build(self) -> Result<(Account, Word), AccountBuildError> {
```

Comment:
There is one other condition we should enforce here: we should not be able to initialize a new account with non-empty vault. Specifically, if `nonce == ZERO`, the user should not be able to add assets to the account (i.e., all account start out with an empty vault).

Reply:
Moved the `nonce` and `with_assets` function behind `testing` but added validation for the non-empty vault anyway so we will notice such issues in testing code as well.